Deploy Environment Weights

joakime opened the pull request:
Enhancement to handle Default Environment sorting via weights assigned to the Environment when the environment is configured with the DeploymentScanner.

joakime committed:
Deploy Environment Weights

joakime committed:
Dump server log if failed to start

joakime committed:
Fix default environment calc

gregw reviewed (changes requested):
I  don't really like the approach of setting weights everywhere.
I'd prefer a centralized approach.  Just have a setter on the scanner that takes a list of environment names and the weight is the index of the env name in that list.   We can have a default list with our known environments and it will be an error if an unknown env name needs to be sorted

gregw reviewed (changes requested):
I don't really like the approach of setting weights everywhere.
I'd prefer a centralized approach. Just have a setter on the scanner that takes a list of environment names and the weight is the index of the env name in that list. We can have a default list with our known environments and it will be an error if an unknown env name needs to be sorted

joakime committed:
Revert XML changes for weight

joakime committed:
Revert EE Test case changes for weight

joakime committed:
Include default weights for known / common environments, but leave custom weight option alone.

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into fix/12.1.x/deploy-environment-weights

joakime commented:
> I don't really like the approach of setting weights everywhere.

I reverted the XML and Test case changes.

> I'd prefer a centralized approach. Just have a setter on the scanner that takes a list of environment names and the weight is the index of the env name in that list. We can have a default list with our known environments and it will be an error if an unknown env name needs to be sorted

I'm good up to the "and it will be an error if an unknown env name needs to be sorted" bit.
I left the new `configureEnvironment(String name, int weight)` alone.
But restored the original `configureEnvironment(String name)`.
Also introduced a properties file that has "our" weights.

This allows for new environments, which your proposal doesn't allow.
It also allows groups to insert their own implementations of ee# environments anywhere they want (even overriding the environments we ship with a weight higher than ours if they want to)

We are down from 24 files changed, to 4 now in this PR.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -131,40 +129,13 @@ public class DeploymentScanner extends ContainerLifeCycle implements Scanner.Bul
     private final Map<String, PathsApp> trackedApps = new HashMap<>();
     private final Map<String, Attributes> environmentAttributesMap = new HashMap<>();
-    private Set<String> trackedEnvironments = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+    private List<TrackedEnv> trackedEnvironments = new ArrayList<>();
```
Make it `final`.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -301,20 +272,56 @@ void addScannerListener(Scanner.Listener listener)
 
+    protected int getDefaultWeight(String name)
+    {
+        URL url = getClass().getResource("environment-weights.properties");
```
I don't think this will work if `DeploymentScanner` is subclassed.

gregw replied:
A properties file is too hard to change and it still makes weights in the "API".   We don't need weights, we just need an order.  If somebody defines a new environment, then setting a new order is easier than inventing a weight that is robust to future changes.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -301,20 +272,56 @@ void addScannerListener(Scanner.Listener listener)
         scanner.addListener(listener);
     }
 
+    protected int getDefaultWeight(String name)
```
I don't like this method.
We should just have `configureEnvironment(String, int)`.

joakime replied:
Read Greg's review, he didn't like this approach

sbordet replied:
Well, I don't like the current one.

We decided to shy away from "default" environments, so the `core` environment was not created upfront, etc. so why now we should have upfront weights?

The weights are not everywhere, they are just in 1 method of `DeploymentScanner`. 
Only Jetty XML code should call that method, so a custom `corp-deploy` module should just be aware of Jetty's `xyz-deploy` modules, if they are used simultaneously with `corp-deploy`.

Otherwise, there are no Jetty `xyz-deploy` modules in use, we don't load our `environment-weights.properties` that we don't use, and there only will be `corp-deploy` and `spring-corp-deploy`, with the weights they decide to use.

joakime replied:
A prior commit did this 
See the revert at https://github.com/jetty/jetty.project/commit/00f95a732032e11ec18a83760adecbd223cb04fd for how it looked.

gregw replied:
I still don't like this approach either. I don't like the concept  of weights being seen externally at all, as it is just an abstraction that we don't need to both the user with.   I think we should just have:
```java
    void setEnvironmentOrder(List<String> environmentNames);
    List<String> getEnvironmentOrder();
```
With the default value being `["core", "static", "ee8", "ee9", "ee10", "ee11"]`. 

If internally we turn that into weights, then fine, but no weights should be in any public API.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -301,20 +272,56 @@ void addScannerListener(Scanner.Listener listener)
+     *    (Example usage: {@code static} is 100, {@code core} is 200, {@code ee8} is 1008, {@code ee11} is 1011)
+     * @return the deployment configuration for the {@link Environment}.
+     */
+    public EnvironmentConfig configureEnvironment(String name, int weight)
```
Just have this method and delete `configureEnvironment(String name)`.
Rename this to `trackEnvironment()` maybe?
The XML files will pass in the weight and there will be no need for the `environment-weights.properties` file.
This method is only called by tests or XML files.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -346,12 +353,29 @@ public String getDefaultEnvironmentName()
+        trackedEnvironments.add(new TrackedEnv(name, weight));
+    }
+
     public void setDefaultEnvironmentName(String name)
```
This method should be removed there is no concept of default, if not determined by the environments that are enabled.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -346,12 +353,29 @@ public String getDefaultEnvironmentName()
+            .toList();
+    }
+
+    protected void addTrackedEnvironment(String name, int weight)
```
Make it private or package.

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -346,12 +353,29 @@ public String getDefaultEnvironmentName()
         {
             return trackedEnvironments.stream()
-                .min(ENVIRONMENT_COMPARATOR)
+                .max(Comparator.comparing(TrackedEnv::weight))
```
Method `getDefaultEnvironmentName()` should be private.

sbordet commented on jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentScannerTest.java:
```diff
@@ -370,23 +371,37 @@ public void testActionListXmlAndWarWithXmlRemoved() throws IOException
+        DeploymentScanner deploymentScanner = new DeploymentScanner(new Server());
+        for (Map.Entry<String, Integer> entry : input.entrySet())
+        {
+            deploymentScanner.addTrackedEnvironment(entry.getKey(), entry.getValue());
```
Should be:

```suggestion
            deploymentScanner.configureEnvironment(entry.getKey(), entry.getValue());
```

sbordet commented on jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java:
```diff
@@ -1684,4 +1721,8 @@ else if (ret != PathsApp.State.REMOVED)
         }
     }
+
+    record TrackedEnv(String name, int weight)
```
TrackedEnv -> TrackedEnvironment

sbordet commented on jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/DeploymentScannerTest.java:
```diff
@@ -370,23 +371,37 @@ public void testActionListXmlAndWarWithXmlRemoved() throws IOException
+    public void testDefaultWeights()
+    {
+        DeploymentScanner deploymentScanner = new DeploymentScanner(new Server());
+        assertThat(deploymentScanner.getDefaultWeight("core"), is(200));
```
There is too much magic here.
Where the 200 come from?

I would expect this test to call `configureEnvironment("core", 200)` before the asserts, so it's clear where the 200 comes from.

gregw reviewed (changes requested):
I don't like weights being in the API at all.  Just have an order list.

sbordet commented:
@gregw I don't see how the "order list" would work in case of custom environments.

With weights in the APIs, users will be exposed to the fact that there is an order, while calling `setEnvironmentOrder()` would be easily missed.

Furthermore, we would have to remember to update that list for every new EE version, while with weights, we just need to copy `jetty-ee11-deploy.xml` into `jetty-ee12-deploy.xml`, and change the new XML file (which we have to anyway, as we likely need to change package names to refer to ee12).

We don't bother users with weights, just implementers of new deployment environments, which should be rare.

sbordet committed:
Removed weights, and implemented the deploy module ordering via [before] directives in the modules.

Removed Environment.CORE constant.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-ee10/jetty-ee10-webapp/src/main/config/modules/ee10-deploy.mod:
```diff
@@ -9,6 +9,12 @@ deployment
 ee10
 
+[before]
+ee9-deploy
+ee8-deploy
+core-deploy
+static-deploy
```
Are you sure you need to list them all?  Will the ordering work if some modules in the order are not enabled?

sbordet replied:
Pretty sure I need to list them all.
When enabling `ee11-deploy` and `static-deploy`, without the full list there is no way to know if `ee11-deploy` is _before_ or not.

sbordet replied:
I just tried and confirmed that the whole list is necessary.

gregw commented on jetty-core/jetty-deploy/src/main/resources/org/eclipse/jetty/deploy/environment-weights.properties:
```diff
@@ -0,0 +1,6 @@
+static=100
+core=200
+ee8=800
+ee9=900
+ee10=1000
+ee11=1100
```
delete this file

gregw reviewed (changes requested):
I'm liking this approach

sbordet committed:
Removed leftover file.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed test assertion.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Restored comment in pom.xml.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed CoreContextHandler ClassLoader initialization.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/12.1.x/deploy-environment-weights'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.1.x' into 'fix/12.1.x/deploy-environment-weights'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
